src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -2696,7 +2743,118 @@ void HwEmShim::closemMessengerThread() {
 		mMessengerThread.join();
 		mMessengerThreadStarted = false;
 	}
+    if(mHostMemAccessThreadStarted) {
+      mHostMemAccessThread.join(); 
+      mHostMemAccessThreadStarted = false;
+    }
 }
 /********************************************** QDMA APIs IMPLEMENTATION END**********************************************/
 /**********************************************HAL2 API's END HERE **********************************************/
+
+/********************************************** Q2H_helper class implementation starts **********************************************/
+bool device2xrt_trans_cb(bool, int, unsigned long int, void*,unsigned long int) {
+    return true;
+}
+Q2H_helper :: Q2H_helper() : buf_size(0),buf(NULL), r(0) {
+    ci_msg = new call_packet_info;
+    ri_msg = new response_packet_info;
+    Q2h_sock = new unix_socket("xcl_sock_K2H");
+    ci_msg->set_size(0);
+    ci_msg->set_xcl_api(0);
+    ri_msg->set_size(0);
+    ri_msg->set_xcl_api(0);
+    i_len = ci_msg->ByteSize();
+    i_buf = malloc(i_len);
+    ri_len = ri_msg->ByteSize();
+    ri_buf = malloc(i_len);
+}
+Q2H_helper::~Q2H_helper() {
+    if(buf != NULL)
+        free(buf);
+    free(i_buf);
+    free(ri_buf);
+    delete ci_msg;
+    delete ri_msg;
+    delete Q2h_sock;
```

Comment:
All this would be unnecessary if the resources were managed by an object (smart pointer)

Reply:
Now unique_ptr is used to manage the dynamic allocation and de-allocation of objects.